File: src/task6_plot_fused_trajectory.py
#!/usr/bin/env python3
"""Plot fused IMU/GNSS trajectories against ground truth for Task 6.

Usage:
    python task6_plot_fused_trajectory.py

This script loads the fused estimator output produced in Task 5 together with
matching ground truth and generates overlay plots in both the NED and ECEF
frames.  Legends are added to all subplots and a quaternion comparison plot is
created across TRIAD, SVD and Davenport methods. Plots are written to the
``results/`` directory.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Dict, Optional

import matplotlib.pyplot as plt
import numpy as np
import scipy.io as sio
from velocity_utils import derive_velocity


# ---------------------------------------------------------------------------
# helper functions
# ---------------------------------------------------------------------------

def ned_to_ecef(
    pos_ned: np.ndarray,
    vel_ned: np.ndarray,
    lat_rad: float,
    lon_rad: float,
) -> tuple[np.ndarray, np.ndarray]:
    """Convert NED position and velocity to the ECEF frame."""
    slat, clat = np.sin(lat_rad), np.cos(lat_rad)
    slon, clon = np.sin(lon_rad), np.cos(lon_rad)
    r_e2n = np.array(
        [
            [-slat * clon, -slat * slon, clat],
            [-slon, clon, 0.0],
            [-clat * clon, -clat * slon, -slat],
        ]
    )
    r_n2e = r_e2n.T
    pos_ecef = pos_ned @ r_n2e
    vel_ecef = vel_ned @ r_n2e
    return pos_ecef, vel_ecef




# ---------------------------------------------------------------------------
# plotting
# ---------------------------------------------------------------------------

def plot_task6_fused_trajectory(
    method: str,
    imu_file: str,
    gnss_file: str,
    quat_logs: Optional[Dict[str, np.ndarray]] = None,
) -> Dict[str, np.ndarray]:
    """Plot fused trajectory for *method* and return quaternion logs."""
    fused_path = Path("results") / f"{imu_file}_{gnss_file}_{method}.mat"
    if not fused_path.exists():
        print(f"Error: missing fused data {fused_path}")
        return quat_logs or {}
    truth_path = Path("results") / f"{imu_file}_{gnss_file}_truth.mat"
    if not truth_path.exists():
        print(f"Error: missing truth data {truth_path}")
        return quat_logs or {}

    fused_data = sio.loadmat(fused_path)
    truth_data = sio.loadmat(truth_path)

    try:
        fused_pos_ned = fused_data["fused_pos"]
        fused_vel_ned = fused_data["fused_vel"]
        fused_time = fused_data["time_s"].squeeze()
        ref_lat = float(np.squeeze(fused_data["ref_lat_rad"]))
        ref_lon = float(np.squeeze(fused_data["ref_lon_rad"]))
        quat_log = fused_data.get("quat_log")
    except KeyError as exc:
        print(f"KeyError: {exc} in {fused_path}")
        return quat_logs or {}

    truth_pos_ned = truth_data["pos_ned_m"]
    truth_time = truth_data["time_s"].squeeze()

    if len(truth_time) != len(fused_time) or not np.allclose(truth_time, fused_time):
        truth_pos_ned = np.vstack(
            [np.interp(fused_time, truth_time, truth_pos_ned[:, i]) for i in range(3)]
        ).T
        truth_time = fused_time

    if "vel_ned_ms" in truth_data:
        truth_vel_ned = truth_data["vel_ned_ms"]
        if len(truth_vel_ned) != len(fused_time):
            truth_vel_ned = np.vstack(
                [np.interp(fused_time, truth_time, truth_vel_ned[:, i]) for i in range(3)]
            ).T
    else:
        print("Warning: deriving truth velocity from position")
        truth_vel_ned = derive_velocity(truth_time, truth_pos_ned)

    fused_pos_ecef, fused_vel_ecef = ned_to_ecef(
        fused_pos_ned, fused_vel_ned, ref_lat, ref_lon
    )
    truth_pos_ecef, truth_vel_ecef = ned_to_ecef(
        truth_pos_ned, truth_vel_ned, ref_lat, ref_lon
    )

    if quat_logs is not None and quat_log is not None:
        quat_logs[method] = quat_log

    error_pos = fused_pos_ned - truth_pos_ned
    error_vel = fused_vel_ned - truth_vel_ned
    rmse_pos = np.sqrt(np.mean(np.sum(error_pos**2, axis=1)))
    rmse_vel = np.sqrt(np.mean(np.sum(error_vel**2, axis=1)))
    final_err = float(np.linalg.norm(error_pos[-1]))

    for frame_name, (p_f, v_f, p_t, v_t) in {
        "NED": (fused_pos_ned, fused_vel_ned, truth_pos_ned, truth_vel_ned),
        "ECEF": (fused_pos_ecef, fused_vel_ecef, truth_pos_ecef, truth_vel_ecef),
    }.items():
        plt.figure(figsize=(10, 8))
        for i, lbl in enumerate(["X", "Y", "Z"] if frame_name == "ECEF" else ["North", "East", "Down"]):
            plt.subplot(3, 1, i + 1)
            plt.plot(fused_time, p_f[:, i], label=f"{method} Fused", color="blue")
            plt.plot(fused_time, p_t[:, i], label="Truth", color="red", linestyle="--")
            plt.title(f"{method} Position {lbl} ({frame_name}, m)")
            plt.xlabel("Time (s)")
            plt.ylabel("Position (m)")
            plt.legend()
            plt.grid(True)
        plt.tight_layout()
        out_name = (
            f"{imu_file}_{gnss_file}_{method}_task6_fused_position_{frame_name.lower()}"
        )
        plt.savefig(Path("results") / f"{out_name}.pdf")
        plt.close()

        plt.figure(figsize=(10, 8))
        for i, lbl in enumerate(["X", "Y", "Z"] if frame_name == "ECEF" else ["North", "East", "Down"]):
            plt.subplot(3, 1, i + 1)
            plt.plot(fused_time, v_f[:, i], label=f"{method} Fused", color="blue")
            plt.plot(fused_time, v_t[:, i], label="Truth", color="red", linestyle="--")
            plt.title(f"{method} Velocity {lbl} ({frame_name}, m/s)")
            plt.xlabel("Time (s)")
            plt.ylabel("Velocity (m/s)")
            plt.legend()
            plt.grid(True)
        plt.tight_layout()
        out_name = (
            f"{imu_file}_{gnss_file}_{method}_task6_fused_velocity_{frame_name.lower()}"
        )
        plt.savefig(Path("results") / f"{out_name}.pdf")
        plt.close()

    plt.figure(figsize=(10, 8))
    for i, lbl in enumerate(["North", "East", "Down"]):
        plt.subplot(3, 1, i + 1)
        plt.plot(fused_time, error_pos[:, i], label=f"{method} Error", color="green")
        plt.title(f"{method} Position Error {lbl} (NED, m)")
        plt.xlabel("Time (s)")
        plt.ylabel("Error (m)")
        plt.legend()
        plt.grid(True)
    plt.tight_layout()
    out_name = f"{imu_file}_{gnss_file}_{method}_task6_position_error_ned"
    plt.savefig(Path("results") / f"{out_name}.pdf")
    plt.close()

    print(
        f"Task 6: {method} final position error {final_err:.3f} m, "
        f"RMSEpos {rmse_pos:.3f} m, RMSEvel {rmse_vel:.3f} m/s"
    )
    return quat_logs or {}


def plot_quaternion_comparison(
    imu_file: str, gnss_file: str, quat_logs: Dict[str, np.ndarray]
) -> None:
    """Plot quaternion components for all methods."""
    if not quat_logs:
        print("No quaternion logs available for comparison")
        return
    plt.figure(figsize=(10, 8))
    labels = ["qw", "qx", "qy", "qz"]
    colors = ["blue", "red", "green", "purple"]
    for i, (lbl, col) in enumerate(zip(labels, colors)):
        plt.subplot(4, 1, i + 1)
        for method, q in quat_logs.items():
            time_s = sio.loadmat(
                Path("results") / f"{imu_file}_{gnss_file}_{method}.mat"
            )["time_s"].squeeze()
            plt.plot(time_s, q[:, i], label=method)
        plt.title(f"Quaternion Component {lbl}")
        plt.xlabel("Time (s)")
        plt.ylabel(lbl)
        plt.legend()
        plt.grid(True)
    plt.tight_layout()
    out_path = Path("results") / f"{imu_file}_{gnss_file}_task6_quaternion_comparison.pdf"
    plt.savefig(out_path)
    plt.close()
    print(f"Saved quaternion comparison to {out_path}")


# ---------------------------------------------------------------------------
# main entry point
# ---------------------------------------------------------------------------


def main() -> None:
    datasets = [("IMU_X002", "GNSS_X002"), ("IMU_X001", "GNSS_X001")]
    methods = ["TRIAD", "SVD", "Davenport"]
    for imu_file, gnss_file in datasets:
        quat_logs: Dict[str, np.ndarray] = {}
        for method in methods:
            print(f"Processing {imu_file}_{gnss_file}_{method}")
            quat_logs = plot_task6_fused_trajectory(method, imu_file, gnss_file, quat_logs)
        plot_quaternion_comparison(imu_file, gnss_file, quat_logs)


if __name__ == "__main__":
    os.makedirs("results", exist_ok=True)
    main()

File: MATLAB/task6_plot_fused_trajectory.m
function quat_logs = task6_plot_fused_trajectory(method, imu_file, gnss_file, quat_logs)
%TASK6_PLOT_FUSED_TRAJECTORY  Plot fused trajectory against truth.
%
%   quat_logs = TASK6_PLOT_FUSED_TRAJECTORY(method, imu_file, gnss_file, quat_logs)
%   replicates ``plot_task6_fused_trajectory`` from the Python code. The fused
%   estimate ``<imu_file>_<gnss_file>_<method>.mat`` and corresponding truth
%   ``<imu_file>_<gnss_file>_truth.mat`` must exist under ``output_matlab/``. Overlay
%   plots in NED and ECEF are produced together with a position error plot. The
%   quaternion history for the method is returned in ``quat_logs``.

if nargin < 4 || isempty(quat_logs)
    quat_logs = struct();
end

fused_path = fullfile('output_matlab', sprintf('%s_%s_%s.mat', imu_file, gnss_file, method));
truth_path = fullfile('output_matlab', sprintf('%s_%s_truth.mat', imu_file, gnss_file));

if ~isfile(fused_path)
    warning('Missing fused data %s', fused_path);
    return;
end
if ~isfile(truth_path)
    warning('Missing truth data %s', truth_path);
    return;
end

F = load(fused_path);
T = load(truth_path);

try
    fused_pos_ned = F.fused_pos;
    fused_vel_ned = F.fused_vel;
    fused_time = F.time_s(:);
    ref_lat = double(F.ref_lat_rad);
    ref_lon = double(F.ref_lon_rad);
catch ME
    warning('KeyError: %s in %s', ME.message, fused_path);
    return;
end

if isfield(F, 'quat_log') && ~isempty(F.quat_log)
    quat_logs.(method) = F.quat_log;
end

truth_pos_ned = T.pos_ned_m;
truth_time = T.time_s(:);

if numel(truth_time) ~= numel(fused_time) || any(abs(truth_time - fused_time) > eps)
    truth_pos_ned = interp1(truth_time, truth_pos_ned, fused_time, 'linear', 'extrap');
    truth_time = fused_time;
end

if isfield(T, 'vel_ned_ms')
    truth_vel_ned = T.vel_ned_ms;
    if size(truth_vel_ned,1) ~= numel(fused_time)
        truth_vel_ned = interp1(T.time_s, truth_vel_ned, fused_time, 'linear', 'extrap');
    end
else
    warning('derive truth velocity from position');
    truth_vel_ned = derive_velocity(truth_time, truth_pos_ned);
end

[fused_pos_ecef, fused_vel_ecef] = ned_to_ecef(fused_pos_ned, fused_vel_ned, ref_lat, ref_lon);
[truth_pos_ecef, truth_vel_ecef] = ned_to_ecef(truth_pos_ned, truth_vel_ned, ref_lat, ref_lon);

error_pos = fused_pos_ned - truth_pos_ned;
error_vel = fused_vel_ned - truth_vel_ned;
rmse_pos = sqrt(mean(sum(error_pos.^2, 2)));
rmse_vel = sqrt(mean(sum(error_vel.^2, 2)));
final_err = norm(error_pos(end, :));

for frame_name = {'NED','ECEF'}
    fr = frame_name{1};
    if strcmp(fr,'NED')
        p_f = fused_pos_ned; v_f = fused_vel_ned;
        p_t = truth_pos_ned; v_t = truth_vel_ned;
        labels = {'North','East','Down'};
    else
        p_f = fused_pos_ecef; v_f = fused_vel_ecef;
        p_t = truth_pos_ecef; v_t = truth_vel_ecef;
        labels = {'X','Y','Z'};
    end

    f1 = figure('Visible','off','Position',[100 100 600 700]);
    for i = 1:3
        subplot(3,1,i); hold on;
        plot(fused_time, p_f(:,i), 'b', 'DisplayName', sprintf('%s Fused', method));
        plot(fused_time, p_t(:,i), 'r--', 'DisplayName', 'Truth');
        title(sprintf('%s Position %s (%s)', method, labels{i}, fr));
        xlabel('Time (s)'); ylabel('Position (m)'); grid on; legend;
    end
    tightfig();
    out_name = sprintf('%s_%s_%s_task6_fused_position_%s', imu_file, gnss_file, method, lower(fr));
    print(f1, fullfile('output_matlab',[out_name '.pdf']), '-dpdf', '-bestfit');
    close(f1);

    f2 = figure('Visible','off','Position',[100 100 600 700]);
    for i = 1:3
        subplot(3,1,i); hold on;
        plot(fused_time, v_f(:,i), 'b', 'DisplayName', sprintf('%s Fused', method));
        plot(fused_time, v_t(:,i), 'r--', 'DisplayName', 'Truth');
        title(sprintf('%s Velocity %s (%s)', method, labels{i}, fr));
        xlabel('Time (s)'); ylabel('Velocity (m/s)'); grid on; legend;
    end
    tightfig();
    out_name = sprintf('%s_%s_%s_task6_fused_velocity_%s', imu_file, gnss_file, method, lower(fr));
    print(f2, fullfile('output_matlab',[out_name '.pdf']), '-dpdf', '-bestfit');
    close(f2);
end

f3 = figure('Visible','off','Position',[100 100 600 700]);
labels_ned = {'North','East','Down'};
for i = 1:3
    subplot(3,1,i); hold on;
    plot(fused_time, error_pos(:,i), 'g', 'DisplayName', sprintf('%s Error', method));
    title(sprintf('%s Position Error %s (NED)', method, labels_ned{i}));
    xlabel('Time (s)'); ylabel('Error (m)'); grid on; legend;
end
tightfig();
out_name = sprintf('%s_%s_%s_task6_position_error_ned', imu_file, gnss_file, method);
print(f3, fullfile('output_matlab',[out_name '.pdf']), '-dpdf', '-bestfit');
close(f3);

fprintf('Task 6: %s final position error %.3f m, RMSEpos %.3f m, RMSEvel %.3f m/s\n', ...
    method, final_err, rmse_pos, rmse_vel);

end

% -------------------------------------------------------------------------
function [pos_ecef, vel_ecef] = ned_to_ecef(pos_ned, vel_ned, lat_rad, lon_rad)
%NED_TO_ECEF Convert NED coordinates to ECEF frame.

slat = sin(lat_rad); clat = cos(lat_rad);
slon = sin(lon_rad); clon = cos(lon_rad);
r_e2n = [-slat*clon, -slat*slon, clat; -slon, clon, 0; -clat*clon, -clat*slon, -slat];
r_n2e = r_e2n.';
pos_ecef = pos_ned * r_n2e;
vel_ecef = vel_ned * r_n2e;
end

% -------------------------------------------------------------------------
function tightfig()
%TIGHTFIG Minimal figure trimming helper.
set(gcf,'PaperPositionMode','auto');
end

